File: 20.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
public class Solution {
    public bool IsValid(string s) {
        if(s.Length%2 == 1) return false;
        Stack<char> stack = new Stack<char>();
        foreach(var c in s){
            if('(' == c || '[' == c || '{' == c)
                stack.Push(c);
            else{
                if(stack.Count == 0) return false;
                switch(c){
                    case ')' : 
                        if(stack.Peek() != '(') return false; break;
                    case ']' : 
                        if(stack.Peek() != '[') return false; break;
                    case '}' : 
                        if(stack.Peek() != '{') return false; break;
                }
                stack.Pop();
            }
        }
        return stack.Count == 0;
    }
}
// @lc code=end

File: 20.valid-parentheses.js
/*
 * @lc app=leetcode id=20 lang=javascript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  var stack = [];
  for (var p of s) {
    if (p == "(" || p == "[" || p == "{") stack.push(p);
    else {
      var last = stack.pop();
      if (last == "(" && p == ")") continue;
      else if (last == "[" && p == "]") continue;
      else if (last == "{" && p == "}") continue;
      return false;
    }
  }
  return stack.length == 0;
};
// @lc code=end
